bundles/mapping/mapmodule/service/map.layer.js

```diff
@@ -329,10 +346,222 @@ Oskari.clazz.define('Oskari.mapframework.service.MapLayerService',
+                        if (layerIndex >= 0) {
+                            if (deleteLayer) {
+                                group.groups[j].layers.splice(layerIndex, 1);
+                            } else {
+                                group.groups[j].layers[layerIndex] = newLayerConf;
+                            }
+                        }
+                        // check subgroup subgroups
+                        if (subgroup.groups) {
+                            for (var k = 0; k < subgroup.groups.length; k++) {
+                                var subgroupSubgroup = subgroup.groups[k];
+                                layerIndex = isLayerInGroup(subgroupSubgroup.layers);
+                                if (layerIndex >= 0) {
+                                    if (deleteLayer) {
+                                        group.groups[j].groups[k].layers.splice(layerIndex, 1);
+                                    } else {
+                                        group.groups[j].groups[k].layers[layerIndex] = newLayerConf;
+                                    }
+                                }
+                            }
+                        }
+
+                        if (layerIndex >= 0) {
+                            break;
+                        }
+                    }
+
+                    if (layerIndex >= 0) {
+                        break;
+                    }
+
+                }
+
+
+                // Finally check at layer has all groups in dom
+                newLayerConf.groups.forEach(function(group) {
+                    var groupConf = me.getAllLayerGroups(group.id);
+                    var isInGroup = jQuery.grep(me.getAllLayerGroups(group.id).layers, function(layer) {
+                        return layer.id === newLayerConf.id;
+                    });
+
+                    if (isInGroup.length === 0) {
+                        me.getAllLayerGroups(group.id).layers.push(newLayerConf);
+                    }
+                });
+            }
+
+            // Also check if layer has removed from groups
+            if (deleteLayer) {
+                // find all groups where layer is
+                // check main groups
+                var hasInGroupLayers = function(group) {
+                    var layers = jQuery.grep(group.layers, function(layer) {
+                        return layer.id === layerId;
+                    });
+                    return layers.length > 0;
+                };
+
+                var getLayerIndex = function(layers) {
+                    var returnIndex = -1;
+                    layers.forEach(function(layer, index) {
+                        if (layer.id === layerId) {
+                            returnIndex = index;
+                        }
+                    });
+                    return returnIndex;
+                };
+                me._layerGroups.forEach(function(group, index) {
```

Comment:
This is basically the same loop-construct that's in the insert/update part of the method as well. Implemented with forEach here and for(var i=0...) above. I'm pretty sure this can be written in a more maintainable style without that much duplication.